### PR TITLE
When temperature sensor is triggered, the user needs to be asked to take a new RDT

### LINKED ISSUE #447: When temperature sensor is triggered, the user needs to be asked to take a new RDT
I noticed that we aren't currently doing anything with temperature sensor triggered barcodes, and we are allowing users to continue through to conduct RDTs. This may not be a huge issue until these barcodes are being used in the field so we could wait for the next release.

When the temperature sensor for RDT barcode is triggered "Sensor triggered: yes" the user needs to be told that the RDT has been exposed to high heat.

Maybe something like this: "This RDT has been flagged for high heat exposure and reliability has been compromised. Please use another RDT."

They should then be taken back a screen to barcode capture/manual entry of RDT information screen.

What do you think @pld ?

### PR DESCRIPTION
**Pull request by syedowaisali90**
Hi @vincent-karuri I have created a PR for review.
Close #447

### CODE REVIEW

**Comment by vincent-karuri** [comment]
@syedowaisali90 this should ideally work like the RDT capture form flow for expired RDTs (as specified in the issue https://github.com/onaio/rdt-standard/issues/447#issue-682037728 ). 

i.e. if the product scanned should not be used (in this case because the sensor is triggered), redirect to a page that asks the user to use a different product. No alert dialogs are necessary.

The current step 2 can be used to display this information. Additional logic needs to be added to the `OneScanCovidRDTBarcodeFactory` to redirect to step 2 if the sensor is triggered.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java [hunk 1]
```diff
@@ -39,4 +40,9 @@ protected String getBarcodeValsAsCSV(Intent data) {
     protected String[] splitCSV(String barcodeCSV) {
         return barcodeCSV.split(",");
     }
-}
+
+    @Override
+    protected void sensorTriggered() {
+        ((RDTJsonFormFragmentPresenter) widgetArgs.getFormFragment().getPresenter()).moveToNextStep("step2");
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java:46]: The step to navigate to should be fetched from the `StepStateConfig`.

Also, confirm that step2 has the correct message as specified in the issue.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java [hunk 2]
```diff
@@ -57,6 +63,8 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
 
     protected abstract String[] splitCSV(String barcodeCSV);
 
+    protected void sensorTriggered() {};
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java]: This method is only relevant in the `OneScanBarcodeFactory` that detects the sensor, no need to have it here.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java [hunk 3]
```diff
@@ -33,14 +34,19 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
         final JsonApi jsonApi = (JsonApi) widgetArgs.getContext();
         if (requestCode == BARCODE_REQUEST_CODE && resultCode == RESULT_OK && data != null) {
             try {
-                String barcodeVals = getBarcodeValsAsCSV(data);
-                jsonApi.writeValue(widgetArgs.getStepName(),
-                        widgetArgs.getJsonObject().optString(JsonFormConstants.KEY),
-                        barcodeVals, "", "", "", false);
+                if (data.getBooleanExtra(CovidConstants.IntentKeys.BARCODE_SENSOR_TRIGGER, false)) {
+                    sensorTriggered();
+                }
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java]: These logic should only be in the OneScanBarcodeFactory, see https://github.com/onaio/rdt-standard/pull/463/commits/8f45c0a0fb14c8796e0faa5de009df1aa5fe2b9b#r484807540.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/OneScanActivity.java [hunk 4]
```diff
@@ -91,6 +92,7 @@ private void setResultAndFinish(OneScanHelper.ScanResponse response) {
                 response.lot, response.productId, String.valueOf(response.sensorTriggered),
                 response.status}, ',');
         resultIntent.putExtra(BARCODE_KEY, barcode);
+        resultIntent.putExtra(CovidConstants.IntentKeys.BARCODE_SENSOR_TRIGGER, response.sensorTriggered);
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/OneScanActivity.java:95]: Why is this needed? I think this value can be extracted from the barcode?

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java [hunk 5]
```diff
@@ -33,14 +34,18 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+                            widgetArgs.getJsonObject().optString(JsonFormConstants.KEY),
+                            barcodeVals, "", "", "", false);
 
-                String[] individualVals = splitCSV(barcodeVals);
-                populateRelevantFields(individualVals);
-                moveToNextStep(convertDate(individualVals[1], "YYYY-MM-dd"));
+                    String[] individualVals = splitCSV(barcodeVals);
+                    populateRelevantFields(individualVals);
+                    moveToNextStep(convertDate(individualVals[1], "YYYY-MM-dd"));
+                }
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java:48]: Even when the sensor is triggered, all other steps, like populating relevant fields, should be performed.

I would recommend modifying `moveToNextStep` to add the sensor trigger check.

**Review by vincent-karuri — COMMENTED** [review]
Also, please add a test that verifies correct action is taken when the sensor is triggered.

**Review by vincent-karuri — CHANGES_REQUESTED** [review]
Also, please add a test that verifies correct action is taken when the sensor is triggered.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java (hunk 1 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java:46]: fixed.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java (hunk 2 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java]: fixed.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java (hunk 3 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java]: fixed.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/OneScanActivity.java (hunk 4 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/OneScanActivity.java:95]: fixed.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java (hunk 5 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java:48]: fixed.

File: rdt-open-reader/src/covidIndonesia/assets/json.form/sample-collection-form-original.json [hunk 6]
```diff
@@ -970,5 +970,24 @@
...
+        "key": "lbl_high_heat_exposure",
+        "type": "label",
+        "text_color": "#000000",
+        "has_bg": true,
+        "bg_color": "#ffffff",
+        "center_label": "true",
+        "top_margin": "15dp",
+        "text": "This RDT has been flagged for high heat exposure and reliability has been compromised. Please use another RDT."
+      }
+    ]
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/assets/json.form/sample-collection-form-original.json:991]: I think there's some misunderstanding here https://github.com/onaio/rdt-standard/pull/463#issuecomment-688092525.

There's no need to add an extra step, simply modify step2 to have this information `This RDT has been flagged for high heat exposure and reliability has been compromised. Please use another RDT.`

File: rdt-open-reader/src/covidIndonesia/assets/json.form/sample-collection-form.json [hunk 7]
```diff
@@ -971,5 +971,24 @@
...
+        "type": "label",
+        "text_color": "#000000",
+        "has_bg": true,
+        "bg_color": "#ffffff",
+        "center_label": "true",
+        "top_margin": "15dp",
+        "text": "{{sample-collection-form-original.step11.lbl_high_heat_exposure.text}}"
+      }
+    ]
+  },
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/assets/json.form/sample-collection-form.json:992]: https://github.com/onaio/rdt-standard/pull/463/commits/3be90ebdfadd9191c4b7cdcf2c1ba65db4ad7347#r485397724

File: rdt-open-reader/src/covidIndonesia/assets/step-config/step-state-config.json [hunk 8]
```diff
@@ -18,5 +18,6 @@
   "covid_wbc_page": "step4",
   "covid_support_investigation_complete_page": "step5",
   "covid_scan_sample_for_delivery_page": "step3",
-  "covid_enter_delivery_details_page": "step4"
+  "covid_enter_delivery_details_page": "step4",
+  "sensor_triggered_page": "step11"
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/assets/step-config/step-state-config.json:22]: https://github.com/onaio/rdt-standard/pull/463/commits/3be90ebdfadd9191c4b7cdcf2c1ba65db4ad7347#r485397724

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java [hunk 9]
```diff
@@ -42,7 +54,35 @@ protected String getBarcodeValsAsCSV(Intent data) {
...
+            JsonApi jsonApi = (JsonApi) widgetArgs.getContext();
+            String stepName = widgetArgs.getStepName();
+            JSONObject stepObject = jsonApi.getStep(stepName);
+            JSONArray fields = stepObject.getJSONArray(JsonFormUtils.FIELDS);
+            for (int i = 0; i < fields.length(); i++) {
+                JSONObject fieldObject = fields.getJSONObject(i);
+                if (TEMP_SENSOR.equals(fieldObject.getString(JsonFormUtils.KEY))) {
+                    result = fieldObject.optBoolean(JsonFormUtils.VALUE, false);
+                }
+            }
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java:81]: It should be possible to get the temp sensor information from the barcode https://github.com/onaio/rdt-standard/blob/3be90ebdfadd9191c4b7cdcf2c1ba65db4ad7347/rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java#L46.

Comment (vincent-karuri) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java:81]: https://github.com/onaio/rdt-standard/blob/3be90ebdfadd9191c4b7cdcf2c1ba65db4ad7347/rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java#L65

Comment (vincent-karuri) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java:81]: You can add another method `moveToNextStep(Date expDate, boolean isSensorTriggered)`  in `CovidRDTBarcodeFactory`. This method will check whether sensor is triggered and also call the original `moveToNextStep(Date expDate)`.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Constants.java [hunk 10]
```diff
@@ -119,6 +119,7 @@
         String SCAN_CARESTART_PAGE = "scan_carestart_page";
         String TWENTY_MIN_COUNTDOWN_TIMER_PAGE = "twenty_min_countdown_timer_page";
         String TAKE_IMAGE_OF_RDT_PAGE = "take_image_of_rdt_page";
+        String SENSOR_TRIGGERED_PAGE = "sensor_triggered_page";
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/util/Constants.java:122]: You can use `product_expired_page` instead, no need for another step.

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+public class OneScanCovidRDTBarcodeFactoryTest {
+
+    private OneScanCovidRDTBarcodeFactory oneScanCovidRDTBarcodeFactory;
+
+    private WidgetArgs widgetArgs;
+    private JsonFormActivity jsonFormActivity;
+
+    @Mock
+    private RDTApplication rdtApplication;
+
```
Comment (vincent-karuri) [inline rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java:45]: You can extend the `WidgetFactoryRobolectricTest`, no need to mock these here.

Comment (vincent-karuri) [reply rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java:45]: Also, this test should ideally move to  `testCovidIndonesia/io.ona.rdt.robolectric`.

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    private void mockStaticClasses() {
+        mockStatic(LayoutInflater.class);
+        LayoutInflater layoutInflater = mock(LayoutInflater.class);
+        mockStatic(LayoutInflater.class);
+        RelativeLayout rootLayout = mock(RelativeLayout.class);
+        PowerMockito.when(LayoutInflater.from(any(Context.class))).thenReturn(layoutInflater);
+        doReturn(rootLayout).when(layoutInflater).inflate(anyInt(), isNull());
+        doReturn(mock(MaterialEditText.class)).when(rootLayout).findViewById(R.id.edit_text);
+        doReturn(mock(Button.class)).when(rootLayout).findViewById(R.id.scan_button);
+    }
```
Comment (vincent-karuri) [inline rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java]: This mocking will not be necessary after https://github.com/onaio/rdt-standard/pull/463/commits/3be90ebdfadd9191c4b7cdcf2c1ba65db4ad7347#r485404727.

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+
+    @After
+    public void tearDown() {
+        oneScanCovidRDTBarcodeFactory = null;
+    }
+
+    @Test
+    public void testMoveToNextStepShouldGotoTwoSensorTriggered() {
+        oneScanCovidRDTBarcodeFactory.moveToNextStep(mock(Date.class));
```
Comment (vincent-karuri) [inline rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java]: There is no `verify` or `assert` statement in the test. Is the correct behaviour really verified in this test?

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.ArgumentMatchers.isNull;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({RDTJsonFormFragment.class, Bundle.class, RDTApplication.class, LayoutInflater.class})
+public class OneScanCovidRDTBarcodeFactoryTest {
```
Comment (vincent-karuri) [inline rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java:34]: This test will undergo quite some refactor when it implements this https://github.com/onaio/rdt-standard/pull/463/commits/3be90ebdfadd9191c4b7cdcf2c1ba65db4ad7347#r485404727.

References to Powermock will need to be stripped out and leave Robolectric and Mockito. Powermock and Robolectric have been known not to play nice together.

File: rdt-open-reader/src/covidIndonesia/resources/sample-collection-form_in.properties [hunk 15]
```diff
@@ -99,3 +99,6 @@ sample-collection-form-original.step3.vtm_brand.label = Merk VTM?
 sample-collection-form-original.step3.vtm_brand.options.lmk.text = LMK
 sample-collection-form-original.step3.doctor_phone_number.v_required.err =
 sample-collection-form-original.step2.lbl_sample_expired.text = Sampel telah kedaluwarsa!
+sample-collection-form-original.step11.title = Pengumpulan sampel
+sample-collection-form-original.step11.next_label = OK
+sample-collection-form-original.step11.lbl_high_heat_exposure.text = RDT ini telah ditandai karena paparan panas tinggi dan keandalannya telah dikompromikan. Harap gunakan RDT lain.
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/resources/sample-collection-form_in.properties:104]: In future, translations to other languages can be left blank. A separate process is used to add them.

File: rdt-open-reader/src/covidIndonesia/assets/json.form/sample-collection-form-original.json (hunk 6 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/assets/json.form/sample-collection-form-original.json:991]: fixed

File: rdt-open-reader/src/covidIndonesia/assets/json.form/sample-collection-form.json (hunk 7 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/assets/json.form/sample-collection-form.json:992]: fixed

File: rdt-open-reader/src/covidIndonesia/assets/step-config/step-state-config.json (hunk 8 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/assets/step-config/step-state-config.json:22]: fixed

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java (hunk 9 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java:81]: i have override `moveToNextStep` method and added a data argument.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Constants.java (hunk 10 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Constants.java:122]: fixed.

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java (hunk 12 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java]: fixed.

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java (hunk 13 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java]: fixed.

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java (hunk 14 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactoryTest.java:34]: fixed.

File: rdt-open-reader/src/covidIndonesia/resources/sample-collection-form_in.properties (hunk 15 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/resources/sample-collection-form_in.properties:104]: translation removed.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java [hunk 16]
```diff
@@ -40,10 +40,8 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
 
                 String[] individualVals = splitCSV(barcodeVals);
                 populateRelevantFields(individualVals);
-                moveToNextStep(convertDate(individualVals[1], "YYYY-MM-dd"));
-            } catch (JSONException e) {
-                Timber.e(e);
-            } catch (ParseException e) {
+                moveToNextStep(data, convertDate(individualVals[1], "YYYY-MM-dd"));
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java:43]: `sensorTriggered` value is already extracted https://github.com/onaio/rdt-standard/blob/9f2f4e819f416fa4b3e5d9b8e6dd0edfbab9ddde/rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java#L65. You can pass that instead.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java [hunk 17]
```diff
@@ -39,4 +48,24 @@ protected String getBarcodeValsAsCSV(Intent data) {
...
+            JsonApi jsonApi = (JsonApi) widgetArgs.getContext();
+            String expiredPageAddr = stepStateConfig.getStepStateObj().optString(Constants.Step.PRODUCT_EXPIRED_PAGE, "step1");
+            try {
+                jsonApi.writeValue(expiredPageAddr, "lbl_sample_expired", "",  "", "", "", false);
+                jsonApi.writeValue(expiredPageAddr, "lbl_collect_new_sample", widgetArgs.getContext().getString(R.string.rdt_high_heat_exposure),  "", "", "", false);
+            } catch (JSONException ex) {
+                Timber.e(ex);
+            }
+            JsonFormFragment nextFragment = RDTJsonFormFragment.getFormFragment(expiredPageAddr);
+            widgetArgs.getFormFragment().transactThis(nextFragment);
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java:66]: This is the same logic applied here https://github.com/onaio/rdt-standard/blob/9c36969c9f51552254872699fe1613d6062af1ac/rdt-open-reader/src/main/java/io/ona/rdt/widget/RDTBarcodeFactory.java#L97.

Why not refactor that logic into a method with signature `moveToNextStep(boolean isUsableProduct)` which moves to next step if product is usable and navigates to `PRODUCT_EXPIRED_PAGE` otherwise.

Then we can call the method here with the value of `sensorTriggered`.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java [hunk 18]
```diff
@@ -39,4 +48,24 @@ protected String getBarcodeValsAsCSV(Intent data) {
...
+    @Override
+    protected void moveToNextStep(Intent data, Date expDate) {
+        final int sensorTriggerIndex = 4;
+        Barcode barcode = data.getParcelableExtra(BARCODE_KEY);
+        if (Boolean.parseBoolean(barcode.displayValue.split(",")[sensorTriggerIndex])) {
+            JsonApi jsonApi = (JsonApi) widgetArgs.getContext();
+            String expiredPageAddr = stepStateConfig.getStepStateObj().optString(Constants.Step.PRODUCT_EXPIRED_PAGE, "step1");
+            try {
+                jsonApi.writeValue(expiredPageAddr, "lbl_sample_expired", "",  "", "", "", false);
+                jsonApi.writeValue(expiredPageAddr, "lbl_collect_new_sample", widgetArgs.getContext().getString(R.string.rdt_high_heat_exposure),  "", "", "", false);
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java:61]: This is not required. The text should be modified within the form step2.

File: rdt-open-reader/src/main/java/io/ona/rdt/widget/MalariaRDTBarcodeFactory.java [hunk 19]
```diff
@@ -47,7 +47,7 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
                     expDate = convertDate(barcodeValues[1].trim(), OPEN_RDT_DATE_FORMAT);
                     populateRelevantFields(barcodeValues, jsonApi, expDate);
                 }
-                moveToNextStep(expDate);
+                moveToNextStep(data, expDate);
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/widget/MalariaRDTBarcodeFactory.java:50]: This change should only affect the `CovidRDTBarcodeFactory` children.

File: rdt-open-reader/src/main/java/io/ona/rdt/widget/RDTBarcodeFactory.java [hunk 20]
```diff
@@ -89,7 +90,7 @@ private boolean isRDTExpired(Date date) {
         return date != null && new Date().after(date);
     }
 
-    protected void moveToNextStep(Date expDate) {
+    protected void moveToNextStep(Intent data, Date expDate) {
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/widget/RDTBarcodeFactory.java:93]: https://github.com/onaio/rdt-standard/pull/463/files#r486818607

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java (hunk 16 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java:43]: how do i get individualVals[4] in `moveToNextStep` method

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java (hunk 18 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java:61]: you mean to add another field with a message in step2 and add the relevance check with step6 temp_sensor.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java (hunk 17 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/OneScanCovidRDTBarcodeFactory.java:66]: you mean to create a new method name `moveToNextStep(boolean isUsableProduct)` inside an `RDTBarcodeFactor.java` class and if so then where can I get the `isUsableProduct` variable from.

**Comment by vincent-karuri** [comment]
Let's discuss the approach here on Monday's standup.

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/widget/OneScanCovidRDTBarcodeFactoryTest.java [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+    public void setUp() throws JSONException {
+        super.setUp();
+        oneScanCovidRDTBarcodeFactory = new OneScanCovidRDTBarcodeFactory();
+        setWidgetArgs();
+        Whitebox.setInternalState(oneScanCovidRDTBarcodeFactory, "stepStateConfig", RDTApplication.getInstance().getStepStateConfiguration());
+    }
+
+    @Test
+    public void testMoveToNextStepShouldGotoTwoSensorTriggered() throws Exception {
+        Whitebox.invokeMethod(oneScanCovidRDTBarcodeFactory, "moveToNextStep", true, Mockito.mock(Date.class));
```
Comment (vincent-karuri) [inline rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/widget/OneScanCovidRDTBarcodeFactoryTest.java:35]: There are no assertions made here.

Could you verify that there was navigation to a different page.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java [hunk 22]
```diff
@@ -41,10 +42,9 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
 
                 String[] individualVals = splitCSV(barcodeVals);
                 populateRelevantFields(individualVals);
-                moveToNextStep(convertDate(individualVals[1], "YYYY-MM-dd"));
-            } catch (JSONException e) {
-                Timber.e(e);
-            } catch (ParseException e) {
+                final int sensorTriggerIndex = 4;
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java:45]: Could you move this to a global constant and share replace this as well  https://github.com/onaio/rdt-standard/blob/a454133b83391cefd15a9f78e6080db225065a79/rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java#L69?

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java [hunk 23]
```diff
@@ -41,10 +42,9 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+                moveToNextStep(Boolean.parseBoolean(individualVals[sensorTriggerIndex]), convertDate(individualVals[1], "YYYY-MM-dd"));
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java:46]: `individualVals[sensorTriggerIndex]`, it is possible for this to be an empty string or null. In this case this should default to true. Please add a private method to perform this logic and invoke the method here.

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/widget/OneScanCovidRDTBarcodeFactoryTest.java [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+    @Before
+    public void setUp() throws JSONException {
+        super.setUp();
+        oneScanCovidRDTBarcodeFactory = new OneScanCovidRDTBarcodeFactory();
+        setWidgetArgs();
+        Whitebox.setInternalState(oneScanCovidRDTBarcodeFactory, "stepStateConfig", RDTApplication.getInstance().getStepStateConfiguration());
+    }
+
+    @Test
+    public void testMoveToNextStepShouldGotoTwoSensorTriggered() throws Exception {
```
Comment (vincent-karuri) [inline rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/widget/OneScanCovidRDTBarcodeFactoryTest.java:34]: Please rename the test to `testMoveToNextStepShouldNavigateToUnusableProductPageIfSensorIsTriggered`

File: rdt-open-reader/src/main/java/io/ona/rdt/widget/RDTBarcodeFactory.java [hunk 25]
```diff
@@ -90,13 +90,17 @@ private boolean isRDTExpired(Date date) {
...
             moveToNextStep();
         } else {
-            String expiredPageAddr = stepStateConfig.getStepStateObj().optString(PRODUCT_EXPIRED_PAGE, "step1");
-            JsonFormFragment nextFragment = RDTJsonFormFragment.getFormFragment(expiredPageAddr);
-            formFragment.transactThis(nextFragment);
+            navigateToUnusableProduct();
         }
     }
+
+    protected void navigateToUnusableProduct() {
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/widget/RDTBarcodeFactory.java:100]: Please rename the method to `navigateToUnusableProductPage`.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java (hunk 22 above)
Comment (vincent-karuri) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java:45]: Also move the `individualVals[1]` index to a global constant and replace this as well https://github.com/onaio/rdt-standard/blob/a454133b83391cefd15a9f78e6080db225065a79/rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java#L66

Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java:45]: fixed.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java (hunk 23 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/widget/CovidRDTBarcodeFactory.java:46]: fixed.

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/widget/OneScanCovidRDTBarcodeFactoryTest.java (hunk 24 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/widget/OneScanCovidRDTBarcodeFactoryTest.java:34]: fixed.

File: rdt-open-reader/src/main/java/io/ona/rdt/widget/RDTBarcodeFactory.java (hunk 25 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/widget/RDTBarcodeFactory.java:100]: fixed.

File: rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/widget/OneScanCovidRDTBarcodeFactoryTest.java (hunk 21 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/testCovidIndonesia/java/io/ona/rdt/robolectric/widget/OneScanCovidRDTBarcodeFactoryTest.java:35]: fixed.

**Review by vincent-karuri — APPROVED** [review]
